src/core/namespace.c

```diff
@@ -469,6 +469,32 @@ static void drop_outside_root(const char *root_directory, MountEntry *m, unsigne
         *n = t - m;
 }
 
+static void mount_mkdir_apivfs(const char *root_directory, MountEntry *m)
+{
+        const char *p, *path;
+
+        assert(m);
+
+        /*
+         * In case RootDirectory= or RootImage= are set, create the /proc, /sys
+         * and /dev directories, it is more likely to be missing when the
+         * service is started under these options. This is consistent with
+         * mount units creating the mount points when missing.
+         */
+        if (!root_directory)
+                return;
+
+        path = mount_entry_path(m);
+        p = startswith(path, root_directory);
+
+        /*
+         * Compare by paths instead of modes since we may bind mount or make
+         * paths READONLY which may hide the original/default mount mode
+         */
+        if (PATH_IN_SET(p, "/proc", "/dev", "/sys"))
+                (void) mkdir_p_label(path, 0755);
```

Comment:
hmm, so if /proc is a symlink then we'd follow that here, but relative to the host's root. I think we should first follow any symlinks with `chase_symlinks(mount_entry_path(m), root_directory, CHASE_NONEXISTENT, &p)`, so that we don't trip on that.

Yes, this is a more theoretic issue (who would actually make /sys and /proc a symlink?) but I think for new code we really should try to be correct on this.